### PR TITLE
Handle HttpResponseException more safely for federated groups

### PR DESCRIPTION
**Pull request by turt2live**
This doesn't feel right, but I think it's in the right direction? This fixes https://github.com/vector-im/riot-web/issues/7204 although it does so by trapping all 4xx errors and re-throwing them for synapse to handle differently, specifically on the group API. 

Riot already has the required bits for handling the 403, but synapse currently returns a 500 instead. See https://github.com/vector-im/riot-web/issues/7204 for details.

Sytest: https://github.com/matrix-org/sytest/pull/512

### CODE REVIEW

File: synapse/handlers/groups_local.py [hunk 1]
```diff
@@ -37,9 +37,17 @@ def f(self, group_id, *args, **kwargs):
...
-            return getattr(self.transport_client, func_name)(
+            d = getattr(self.transport_client, func_name)(
                 destination, group_id, *args, **kwargs
             )
+            def h(failure):
+                failure.trap(HttpResponseException)
+                e = failure.value
+                if e.code >= 400 and e.code < 500:
+                    raise SynapseError(e.code, e.msg)
+                failure.raiseException()
```
Comment (richvdh) [inline synapse/handlers/groups_local.py]: I think you can just `return failure` here.

File: synapse/handlers/groups_local.py [hunk 2]
```diff
@@ -37,9 +37,17 @@ def f(self, group_id, *args, **kwargs):
...
             destination = get_domain_from_id(group_id)
-            return getattr(self.transport_client, func_name)(
+            d = getattr(self.transport_client, func_name)(
                 destination, group_id, *args, **kwargs
             )
+            def h(failure):
+                failure.trap(HttpResponseException)
+                e = failure.value
+                if e.code >= 400 and e.code < 500:
+                    raise SynapseError(e.code, e.msg)
```
Comment (richvdh) [inline synapse/handlers/groups_local.py]: note that `HttpResponseException` has a `to_synapse_error` method which you probably want here.

File: synapse/handlers/groups_local.py [hunk 3]
```diff
@@ -37,9 +37,17 @@ def f(self, group_id, *args, **kwargs):
...
         else:
             destination = get_domain_from_id(group_id)
-            return getattr(self.transport_client, func_name)(
+            d = getattr(self.transport_client, func_name)(
                 destination, group_id, *args, **kwargs
             )
+            def h(failure):
+                failure.trap(HttpResponseException)
+                e = failure.value
+                if e.code >= 400 and e.code < 500:
```
Comment (richvdh) [inline synapse/handlers/groups_local.py]: I think I'd like us to be conservative about which error codes we pass through here; can you stat with `if e.code == 403` and we'll work up from there?

**Comment by richvdh** [comment]
(otherwise it looks sensible to me)

**Comment by richvdh** [comment]
(CI is failing due to pep8/import sorting)

**Comment by turt2live** [comment]
I'll work on sytests next week alongside the other PRs I have.

**Comment by turt2live** [comment]
Have made a sytest: https://github.com/matrix-org/sytest/pull/512

File: synapse/handlers/groups_local.py [hunk 4]
```diff
@@ -37,9 +37,18 @@ def f(self, group_id, *args, **kwargs):
             )
         else:
             destination = get_domain_from_id(group_id)
-            return getattr(self.transport_client, func_name)(
+            d = getattr(self.transport_client, func_name)(
                 destination, group_id, *args, **kwargs
             )
+
+            def h(failure):
```
Comment (richvdh) [inline synapse/handlers/groups_local.py]: a comment about what we're doing here wouldn't go amiss.

Comment (turt2live) [reply synapse/handlers/groups_local.py]: added some longish comments

**Pull request merged**